### DISCUSSION TITLE
How to generate C code from a code containing Opti class?

### DISCUSSION THREAD

**Discussion by AliRasoulzadeh1990** [casadi/casadi#3432] [Q&A] [answered]
Suppose this simple example. I need to solve the simple NLP of a function f = x^2+y^2 subject to constraint g = a*x+y-b = 0, lower/upper bounds for x and an initial x0. With following python code:
```
from casadi import *

# parameters:
a = 1
b = 2
 
opti = Opti()

x = opti.variable()
y = opti.variable()

f = x * x + y * y 
opti.minimize(f)

g = a * x + y - b
opti.subject_to(g == 0)
 
opti.solver('sqpmethod')
solution = opti.solve()

print("x_optimum = ", solution.value(x))
print("y_optimum = ", solution.value(y))
print("f_optimum = ", solution.value(f))
print("g_optimum = ", solution.value(g))
```
So, I want to generate a C code that its main function gets the "a" and "b" parameters as input, and the optimum x & y as the output. But I am confused; since the examples in CasADi documentation show that code generation is done on a casadi function by calling the "codegen" method of the function object. But here, I want to encompass the whole process.

**Comment by jgillis** [comment, marked as answer]
The key is to use opti.to_function

**Reply by AliRasoulzadeh1990** [reply]
Thank you, it worked. But it says "The function "qpsol", which is of type "QpoasesInterface" cannot be code generated. The generation will proceed, but compilation of the code will not be possible."

Which one of the solvers working with CasADi is compilable?

**Reply by jgillis** [reply]
qrqp, osqp and highs iirc

**Reply by AliRasoulzadeh1990** [reply]
Cool! Thank you, I used the qrqp and generated the C code.
Then, I used the following command in terminal for compiling:
```gcc -o optimizer optimizer.c```
Another issue is that when compiling, it says in function 'optimizer_house' and three other functions: "undefined reference to sqrt" which when I change them to casadi_sq, the issue is solved. But another error is that in function 'optimizer_fmin' and 'optimizer_fmax', also "undefined reference to fmin (and fmax)", but they couldn't be changed to anything, because they are in the definition of optimizer_fmin (and _fmax)... in the generated code, "casadi" is addressed instead of my chosen "optimizer" name, as you may know it, and it's okay.

**Reply by AliRasoulzadeh1990** [reply]
I added -lm to the end of compilation command and it is solved. It needed math library.
